src/matlab.jl

```diff
@@ -0,0 +1,274 @@
+
+im_from_matlab(img ./ 255) # convert to Float64
+im_from_matlab(UInt8.(img)) # convert to UInt8
+```
+
+Indexd image in MATLAB convention consists of the `index`-`values` pair. `values` is a
+two-dimensional N×3 numerical array, and `index` is a integer-valued array in range \$[1,
+N]\$.
+
+```julia
+# a 4×4 random indexed image using five colors
+index = rand(1:5, 4, 4)
+values = [0.0 0.0 0.0  # black
+          1.0 0.0 0.0  # red
+          0.0 1.0 0.0  # green
+          0.0 0.0 1.0  # blue
+          1.0 1.0 1.0] # white
+
+# 4×4 matrix with eltype RGB{Float64}
+im_from_matlab(index, values)
+```
+
+!!! tip "eager conversion"
+    To save memory allocation, the conversion is done in lazy mode. In some cases, this
+    could introduce performance overhead due to the repeat computation. This can be easily
+    solved by converting eagerly via, e.g., `collect(im_from_matlab(...))`.
+
+See also: [`im_to_matlab`](@ref).
+"""
+function im_from_matlab end
+
+# Step 1: convenient conventions
+# - 1d numerical vector is Gray image
+# - 2d numerical array is Gray image
+# - 3d numerical array of size (m, n, 3) is RGB image
+# For other cases, users must specify `CT` explicitly; otherwise it is not type-stable
+im_from_matlab(X::AbstractVector) = vec(im_from_matlab(reshape(X, (length(X), 1))))
+im_from_matlab(X::AbstractMatrix{T}) where {T<:Real} = im_from_matlab(Gray, X)
+function im_from_matlab(X::AbstractArray{T,3}) where {T<:Real}
+    if size(X, 3) != 3
+        msg = "Unrecognized MATLAB image layout."
+        hint = size(X, 3) == 1 ? "Do you mean `im_from_matlab(reshape(X, ($(size(X)[1:2]...))))`?" : ""
+        msg = isempty(hint) ? msg : "$msg $hint"
+        throw(ArgumentError(msg))
+    end
+    return im_from_matlab(RGB, X)
+end
+im_from_matlab(X::AbstractArray) = throw(ArgumentError("Unrecognized MATLAB image layout."))
+
+# Step 2: storage type conversion
+function im_from_matlab(::Type{CT}, X::AbstractArray{T}) where {CT,T}
+    if T <: Union{Normed,AbstractFloat}
+        return _im_from_matlab(CT, X)
+    else
+        msg = "Unrecognized element type $T, manual conversion to float point number or fixed point number is needed."
+        hint = _matlab_type_hint(X)
+        msg = isempty(hint) ? msg : "$msg $hint"
+        throw(ArgumentError(msg))
+    end
+end
+im_from_matlab(::Type{CT}, X::AbstractArray{UInt8}) where {CT} = _im_from_matlab(CT, reinterpret(N0f8, X))
+im_from_matlab(::Type{CT}, X::AbstractArray{UInt16}) where {CT} = _im_from_matlab(CT, reinterpret(N0f16, X))
+function im_from_matlab(::Type{CT}, X::AbstractArray{Int16}) where {CT}
+    # MALTAB compat
+    _im2double(x) = (Float64(x) + Float64(32768)) / Float64(65535)
+    return _im_from_matlab(CT, mappedarray(_im2double, X))
+end
+
+function _matlab_type_hint(@nospecialize X)
+    mn, mx = extrema(X)
+    if mn >= typemin(UInt8) && mx <= typemax(UInt8)
+        return "For instance: `UInt8.(X)` or `X./$(typemax(UInt8))`"
```

Comment:
I wonder if we should also check `isinteger(mn) && isinteger(mx)` or even `all(isinteger, X)` before recommending `UInt8.(X)`? (`isinteger` checks the value, not the type, so it may be appropriate for an image that happens to be encoded with `Float64` but for which all values are actually integers.)